src/providers/ogr/qgsogrprovider.cpp

```diff
@@ -1175,7 +1176,8 @@ size_t QgsOgrProvider::layerCount() const
  */
 QgsWkbTypes::Type QgsOgrProvider::wkbType() const
 {
-  return static_cast<QgsWkbTypes::Type>( mOGRGeomType );
+  QgsWkbTypes::Type wkb = static_cast<QgsWkbTypes::Type>( mOGRGeomType );
+  return ogrDriverName == QLatin1String( "ESRI Shapefile" ) ? QgsWkbTypes::multiType( wkb ) : wkb;
```

Comment:
@rouault , OK. I'll update this PR to only do LineString -> MultiLineString and Polygon -> MultiPolygon.

Regarding feature iterator, @nyalldawson might have some good thoughts on that (certainly within the context of the processing framework).